File: DataModule/src/com/mobnetic/coinguardian/model/market/Cryptonit.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;

public class Cryptonit extends Market {

	private final static String NAME = "Cryptonit";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://cryptonit.net/apiv2/rest/public/ccorder.json?bid_currency=%1$s&ask_currency=%2$s&ticker";
	private final static String URL_CURRENCY_PAIRS = "https://cryptonit.net/apiv2/rest/public/pairs.json";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();
	
	public Cryptonit() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyCounterLowerCase(), checkerInfo.getCurrencyBaseLowerCase());		// reversed pairs!
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject rateJsonObject = jsonObject.getJSONObject("rate");
		final JSONObject volumeJsonObject = jsonObject.getJSONObject("volume");
		
		ticker.bid = getDoubleFromString(rateJsonObject, "bid");
		ticker.ask = getDoubleFromString(rateJsonObject, "ask");
		if(volumeJsonObject.has(checkerInfo.getCurrencyBase()))
			ticker.vol = getDoubleFromString(volumeJsonObject, checkerInfo.getCurrencyBase());
		ticker.high = getDoubleFromString(rateJsonObject, "high");
		ticker.low = getDoubleFromString(rateJsonObject, "low");
		ticker.last = getDoubleFromString(rateJsonObject, "last");
	}
	
	private double getDoubleFromString(JSONObject jsonObject, String name) throws NumberFormatException, JSONException {
		return Double.parseDouble(jsonObject.getString(name));
	}
	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}
	
	@Override
	protected void parseCurrencyPairs(int requestId, String responseString, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONArray pairsJsonArray = new JSONArray(responseString);
		JSONArray currenciesJsonArray = null;
		for(int i=0; i<pairsJsonArray.length(); ++i) {
			currenciesJsonArray = pairsJsonArray.getJSONArray(i);
			if(currenciesJsonArray.length()!=2)
				continue;
			final String currencyBase = currenciesJsonArray.getString(1);		// reversed pairs!
			final String currencyCounter = currenciesJsonArray.getString(0);	// reversed pairs!
			if(currencyBase==null || currencyCounter==null)
				continue;
			
			pairs.add(new CurrencyPairInfo(currencyBase.toUpperCase(Locale.US), currencyCounter.toUpperCase(Locale.US), null));
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Cryptonit.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.util.ParseUtils
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

class Cryptonit : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyCounterLowerCase, checkerInfo.currencyBaseLowerCase) // reversed pairs!
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val rateJsonObject = jsonObject.getJSONObject("rate")
        val volumeJsonObject = jsonObject.getJSONObject("volume")
        ticker.bid = ParseUtils.getDoubleFromString(rateJsonObject, "bid")
        ticker.ask = ParseUtils.getDoubleFromString(rateJsonObject, "ask")
        if (volumeJsonObject.has(checkerInfo.currencyBase)) {
            ticker.vol = ParseUtils.getDoubleFromString(volumeJsonObject, checkerInfo.currencyBase)
        }
        ticker.high = ParseUtils.getDoubleFromString(rateJsonObject, "high")
        ticker.low = ParseUtils.getDoubleFromString(rateJsonObject, "low")
        ticker.last = ParseUtils.getDoubleFromString(rateJsonObject, "last")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairs(requestId: Int, responseString: String, pairs: MutableList<CurrencyPairInfo>) {
        val pairsJsonArray = JSONArray(responseString)
        var currenciesJsonArray: JSONArray? = null
        for (i in 0 until pairsJsonArray.length()) {
            currenciesJsonArray = pairsJsonArray.getJSONArray(i)
            if (currenciesJsonArray.length() != 2) continue
            val currencyBase = currenciesJsonArray.getString(1) // reversed pairs!
            val currencyCounter = currenciesJsonArray.getString(0) // reversed pairs!
            if (currencyBase == null || currencyCounter == null) continue
            pairs.add(CurrencyPairInfo(currencyBase.toUpperCase(Locale.US), currencyCounter.toUpperCase(Locale.US), null))
        }
    }

    companion object {
        private const val NAME = "Cryptonit"
        private const val TTS_NAME = NAME
        private const val URL = "https://cryptonit.net/apiv2/rest/public/ccorder.json?bid_currency=%1\$s&ask_currency=%2\$s&ticker"
        private const val URL_CURRENCY_PAIRS = "https://cryptonit.net/apiv2/rest/public/pairs.json"
    }
}